### PR TITLE
sim: Enable garbage collection of unused input sections 

### PR DESCRIPTION
**Pull request by anchao**
## Summary

sim: Enable garbage collection of unused input sections

LDFLAGS += -Wl,--gc-sections

GC should be enabled on arch/sim/src/Makefile:

```
326   $(if $(CONFIG_HAVE_CXX),\
327   $(Q) "$(CXX)" $(CFLAGS) $(LDFLAGS) -o $(TOPDIR)/$@ $(HEADOBJ) nuttx.rel $(HOSTOBJS) $(STDLIBS),\
328   $(Q) "$(CC)" $(CFLAGS) $(LDFLAGS) -o $(TOPDIR)/$@ $(HEADOBJ) nuttx.rel $(HOSTOBJS) $(STDLIBS))
```

## Impact

N/A

## Testing

CI-CHECK

### CODE REVIEW

File: boards/sim/sim/sim/scripts/Make.defs [hunk 1]
```diff
@@ -209,6 +209,7 @@ endif
 
 LDELFFLAGS = -r -e main --gc-sections
 LDELFFLAGS += -T $(call CONVERT_PATH,$(BOARD_DIR)$(DELIM)scripts$(DELIM)gnu-elf.ld)
+LDFLAGS += -Wl,--gc-sections
```
Comment (pkarashchenko) [inline boards/sim/sim/sim/scripts/Make.defs:212]: Do we need to remove `--gc-sections` from `LDELFFLAGS`?

Comment (anchao) [reply boards/sim/sim/sim/scripts/Make.defs:212]: LDFLAGS is different with LDELFFLAGS, LDELFFLAGS used in loadable applications

**Review by pkarashchenko — APPROVED** [review]

File: boards/sim/sim/sim/scripts/Make.defs (hunk 1 above)
Comment (pkarashchenko) [inline boards/sim/sim/sim/scripts/Make.defs:212]: I think that `-Wl,-dead_strip` is needed for MAC

File: boards/sim/sim/sim/scripts/Make.defs [hunk 2]
```diff
@@ -209,6 +209,11 @@ endif
...
+
+ifeq ($(CONFIG_HOST_MACOS),y)
+  LDFLAGS += -Wl,-dead_strip
+endif
```
Comment (pkarashchenko) [inline boards/sim/sim/sim/scripts/Make.defs:216]: Probably gc-sections should be in else case

Comment (anchao) [reply boards/sim/sim/sim/scripts/Make.defs:217]: Done

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**